### PR TITLE
Add omnibus metrics

### PR DESCRIPTION
**Pull request by makyo**

### CODE REVIEW

File: lib/charms/layer/jujushell.py [hunk 1]
```diff
@@ -317,3 +317,7 @@ def exterminate_containers():
         # 2.2.4 we can replace the suprocess call with calls to
         # container.stop(wait=True) and container.delete()
         call('/snap/bin/lxc', 'delete', '-f', container.name)
+
+
+def fetch_metrics():
```
Comment (frankban) [inline lib/charms/layer/jujushell.py]: Leftover?

File: metrics.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+metrics:
+    requests_in_flight:
+        type: gauge
+        description: Number of requests in flight.
+        command: scripts/metrics_scripts/requests_in_flight
```
Comment (frankban) [inline metrics.yaml]: Can't we call the collect_metrics.py with the right argument directly from here, without having to create a script for every metric?

File: scripts/collect_metrics.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    'errors_count': 'jujushell_errors_count',
+    'containers_in_flight': 'jujushell_containers_in_flight',
+    'containers_duration_create_container':
+        'jujushell_containers_duration_sum{operation="create-container"}',
+    'containers_duration_get_all_containers':
+        'jujushell_containers_duration_sum{operation="get-all-containers"}',
+}
+
+
+def get_metric(metric_name):
```
Comment (frankban) [inline scripts/collect_metrics.py:21]: Please add a doctsring.

File: metrics.yaml [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    requests_count:
+        type: gauge
+        description: Count of requests.
+        command: scripts/metrics_scripts/requests_count
+    requests_duration:
+        type: gauge
+        description: The duration of requests.
```
Comment (frankban) [inline metrics.yaml]: In seconds?

File: metrics.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        type: gauge
+        description: The number of encountered errors
+        command: scripts/metrics_scripts/errors_count
+    containers_in_flight:
+        type: gauge
+        description: The number of containers currently present in the unit.
+        command: scripts/metrics_scripts/containers_in_flight
+    containers_duration_create_container:
+        type: gauge
+        description: Time spent creating containers.
```
Comment (frankban) [inline metrics.yaml]: In seconds?

File: metrics.yaml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        command: scripts/metrics_scripts/containers_duration_create_container
+    containers_duration_get_all_containers:
+        type: gauge
+        description: Time spent getting containers.
```
Comment (frankban) [inline metrics.yaml]: Seconds?

File: scripts/collect_metrics.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    metric = METRIC_NAME_MAP.get(metric_name)
+    if metric is None:
+        return ''
+    try:
+        response = request.urlopen('http://127.0.0.1/metrics')
+    except HTTPError:
+        return ''
+    for line in response:
+        if line.startswith(METRIC_NAME_MAP[metric_name]):
```
Comment (frankban) [inline scripts/collect_metrics.py]: s/METRIC_NAME_MAP[metric_name]/metric/

File: scripts/test_collect_metrics.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+from unittest import (
```
Comment (frankban) [inline scripts/test_collect_metrics.py]: I would be nice to have all tests together in the /test directory, and use the sys path stuff already in place there to make collect_metrics importable.

File: scripts/test_collect_metrics.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            'jujushell_containers_duration_bucket{operation="get-all-containers",le="10"} 16',
+            'jujushell_containers_duration_bucket{operation="get-all-containers",le="+Inf"} 16',
+            'jujushell_containers_duration_sum{operation="get-all-containers"} 0.127221213',
+            'jujushell_containers_duration_count{operation="get-all-containers"} 16',
+            '# HELP jujushell_containers_in_flight the number of containers currently present in the machine',
+            '# TYPE jujushell_containers_in_flight gauge',
+            'jujushell_containers_in_flight 2',
+            '# HELP jujushell_errors_count the number of encountered errors',
+            '# TYPE jujushell_errors_count counter',
+            'jujushell_errors_count{message="cannot log into juju: cannot authenticate user: interaction required but not possible"} 1',
```
Comment (frankban) [inline scripts/test_collect_metrics.py]: Some of those lines are generating lint failures.

File: scripts/test_collect_metrics.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            'jujushell_requests_duration_sum{code="200"} 28474.801069908997',
+            'jujushell_requests_duration_count{code="200"} 17',
+            '# HELP jujushell_requests_in_flight the number of requests currently in flight',
+            '# TYPE jujushell_requests_in_flight gauge',
+            'jujushell_requests_in_flight 0',
+            '# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.',
+            '# TYPE process_cpu_seconds_total counter',
+            'process_cpu_seconds_total 292.48',
+        ]
+        self.assertEqual(collect_metrics.get_metric('dalek'), '')
```
Comment (frankban) [inline scripts/test_collect_metrics.py]: I'd create 4 different test methods for:
1. success case,
2. empty metrics due to invalid metric name,
3. empty metrics due to valid metric in the map but not in the response body,
4. http error exception

**Review by frankban — COMMENTED** [review]
This looks good with some comments. 
Was using a prometheus lib for parsing the metrics a bit overkill? 
Thanks!

**Review by jcsackett — APPROVED** [review]
This looks good, although I echo @frankban's hope that we can call directly rather than creating a slew of wrapper scripts.

**Pull request merged**